### PR TITLE
Add `BetaGeoBetaBinomModel`

### PR DESCRIPTION
**Pull request by ColtAllen**
## Description

Reopening https://github.com/pymc-labs/pymc-marketing/pull/922 due to the force push.

The BG/BB model is for non-contractual purchase opportunities across discrete time periods; a good example would be sporting events.

Sampling is rather slow because NUTS defaults to compound sampling due to the discrete distributions used in this model. I recommend that https://github.com/pymc-labs/pymc-marketing/pull/707 be merged to speed up performance. 

A small addition to `CLVBaseModel` was required for the input validations, and the dev notebook is rather minimalist right now but will be expanded into a full how-to in a future PR. An inefficiency in `distribution_new_customers` that is shared by all CLV models was also identified and will be fixed in a separate PR.

## Related Issue

- [ ] Closes https://github.com/pymc-labs/pymc-marketing/issues/176
- [ ] Related to https://github.com/pymc-labs/pymc-marketing/pull/707

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [ ] MMM
- [x] CLV

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc-marketing--1031.org.readthedocs.build/en/1031/

### CODE REVIEW

**Comment by ColtAllen** [comment]
Is there a `ruff` version discrepancy between the CI and the pre-commit config? I'm not getting these errors when working locally and all dependencies are up to date.

File: pymc_marketing/clv/models/__init__.py [hunk 1]
```diff
@@ -16,15 +16,18 @@
...
 from pymc_marketing.clv.models.beta_geo import BetaGeoModel
+from pymc_marketing.clv.models.beta_geo_beta_binom import BetaGeoBetaBinomModel
 from pymc_marketing.clv.models.gamma_gamma import (
     GammaGammaModel,
     GammaGammaModelIndividual,
 )
 from pymc_marketing.clv.models.pareto_nbd import ParetoNBDModel
 from pymc_marketing.clv.models.shifted_beta_geo import ShiftedBetaGeoModelIndividual
 
+
```
Comment (juanitorduz) [inline pymc_marketing/clv/models/__init__.py:27]: Also removing this empty line (also get a complain locally)

**Comment by juanitorduz** [comment]
pre-commit.ci autofix

File: pymc_marketing/clv/models/__init__.py (hunk 1 above)
Comment (juanitorduz) [reply pymc_marketing/clv/models/__init__.py]: Fixed by https://github.com/pymc-labs/pymc-marketing/pull/1031/commits/b3abdab678ac7ae46a5491d26911cada469d494f using https://github.com/pymc-labs/pymc-marketing/pull/1031#issuecomment-2349538095

File: pymc_marketing/clv/__init__.py [hunk 2]
```diff
@@ -15,6 +15,7 @@
 
 from pymc_marketing.clv.models import (
     BetaGeoModel,
+    BetaGeoBetaBinomModel,
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply pymc_marketing/clv/__init__.py:17]: Fixed by https://github.com/pymc-labs/pymc-marketing/pull/1031/commits/b3abdab678ac7ae46a5491d26911cada469d494f using https://github.com/pymc-labs/pymc-marketing/pull/1031#issuecomment-2349538095

**Comment by juanitorduz** [comment]
Fixed by https://github.com/pymc-labs/pymc-marketing/pull/1031/commits/b3abdab678ac7ae46a5491d26911cada469d494f using https://github.com/pymc-labs/pymc-marketing/pull/1031#issuecomment-2349538095 

@ColtAllen If you wanna add some changed make sure to pull first ;)

**Comment by juanitorduz** [comment]
@ColtAllen regarding 

> Sampling is rather slow because NUTS defaults to compound sampling due to the discrete distributions used in this model. I recommend that https://github.com/pymc-labs/pymc-marketing/pull/707 be merged to speed up performance.

Do you see a speed-up with this model or all CLV models?

File: pymc_marketing/clv/models/beta_geo_beta_binom.py [hunk 3]
```diff
@@ -0,0 +1,701 @@
...
+                delta=delta,
+                gamma=gamma,
+                T=self.data["T"],
+                observed=np.stack(
+                    (self.data["recency"], self.data["frequency"]), axis=1
+                ),
+                dims=["customer_id", "obs_var"],
+            )
+
+    # TODO: cache this as a property
```
Comment (juanitorduz) [inline pymc_marketing/clv/models/beta_geo_beta_binom.py:239]: Do you want to create an issue to track this TODO?

File: pymc_marketing/clv/models/beta_geo_beta_binom.py [hunk 4]
```diff
@@ -0,0 +1,701 @@
...
+            dummy_logp,
+            replace={dummy_T: T.values[:, None], dummy_values: values[:, None, :]},
+        ).eval()
+        # Unstack chain/draw and put customer in last axis
+        loglike = np.moveaxis(
+            loglike.reshape((-1, alpha.sizes["chain"], alpha.sizes["draw"])), 0, -1
+        )
+        return xarray.DataArray(data=loglike, dims=("chain", "draw", "customer_id"))
+
+    # TODO: move this into BaseModel
```
Comment (juanitorduz) [inline pymc_marketing/clv/models/beta_geo_beta_binom.py:278]: Also create an issue?

**Review by juanitorduz — APPROVED** [review]
Fantastic work @ColtAllen !

**Comment by juanitorduz** [comment]
@ColtAllen 

This test is very slow:
```
6182.49s call     tests/clv/models/test_beta_geo_beta_binom.py::TestBetaGeoBetaBinomModel::test_model_convergence[mcmc-0.1]
```
Any ideas if we can make it a bit faster?

**Comment by ColtAllen** [comment]
> Do you see a speed-up with this model or all CLV models?

https://github.com/pymc-labs/pymc-marketing/pull/707 only impacts `BetaGeoBetaBinomModel`.

> This test is very slow. Any ideas if we can make it a bit faster?

Merging https://github.com/pymc-labs/pymc-marketing/pull/707 would make it about 3x faster. Only other way to speed it up would be to use a smaller fit dataset. There's another test here that is also rather slow, but addressing the TODOs will fix that.

**Comment by juanitorduz** [comment]
Ok! Then from my side we can merge this one and lets work on those to-dos 🙂

**Pull request merged**